### PR TITLE
Support Aliyun/Tencent/Huawe hosted k8s

### PR DESCRIPTION
**Pull request by loganhz**
Proposed changes
======
Support Aliyun/Tencent/Huawe hosted k8s

Types of changes
======
- New feature (non-breaking change which adds functionality)

Linked Issues
======
https://github.com/rancher/rancher/issues/16809
 

Further comments
======

### CODE REVIEW

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js [hunk 1]
```diff
@@ -0,0 +1,819 @@
...
+
+      set(this, 'config.clusterEipId', defaultId)
+    } else {
+      set(this, 'config.clusterEipId', null)
+    }
+  }),
+
+  clusterFlavorObserver: observer('managementScale', 'highAvailabilityEnabled', function() {
+    const managementScale = get(this, 'managementScale')
+    const highAvailabilityEnabled = get(this, 'highAvailabilityEnabled')
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js]: you can destructure this

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js [hunk 2]
```diff
@@ -0,0 +1,819 @@
...
+        'config.bmsPeriodNum':  null,
+        'config.bmsPeriodType': null,
+      })
+
+      return
+    }
+    const arr = validityPeriod.split(' ')
+
+    set(this, 'config.bmsPeriodNum', parseInt(arr[0]))
+    set(this, 'config.bmsPeriodType', arr[1])
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js]: ```setProperties(this.config, ...```

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js [hunk 3]
```diff
@@ -0,0 +1,819 @@
...
+        'validityPeriod':        '1 month',
+      })
+    }
+  }),
+
+  managementScaleContent: computed('config.clusterType', function() {
+    const clusterType = get(this, 'config.clusterType')
+
+    if (clusterType === 'BareMetal') {
+      return [{
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js]: fix formatting in this and the next array

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js [hunk 4]
```diff
@@ -0,0 +1,819 @@
...
+      value: 'small',
+    }, {
+      label: '200',
+      value: 'medium',
+    }, {
+      label: '1000',
+      value: 'large',
+    }]
+  }),
+  editing: computed('mode', function() {
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js]:
```suggestion
  editing: equal('mode', 'edit')
```

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js [hunk 5]
```diff
@@ -0,0 +1,819 @@
...
+
+    return billingMode === 0 ? intl.t('clusterNew.huaweicce.billingMode.payPerUse') : intl.t('clusterNew.huaweicce.billingMode.yearly')
+  }),
+
+  billingModeContent: computed('config.clusterType', function() {
+    const clusterType = get(this, 'config.clusterType')
+    const intl = get(this, 'intl')
+
+    if (clusterType === 'VirtualMachine') {
+      return [{
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js]: formatting

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js [hunk 6]
```diff
@@ -0,0 +1,819 @@
...
+      }, {
+        label: intl.t('clusterNew.huaweicce.billingMode.yearly'),
+        value: 2,
+      }]
+    }
+  }),
+
+  validityPeriodName: computed('config.bmsPeriodNum', 'config.bmsPeriodType', function() {
+    const bmsPeriodNum = get(this, 'config.bmsPeriodType')
+    const bmsPeriodType = get(this, 'config.bmsPeriodType')
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js]: you can destructure here

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js [hunk 7]
```diff
@@ -0,0 +1,819 @@
...
+
+    return `${ bmsPeriodNum } ${ bmsPeriodType }`
+  }),
+
+  bmsIsAutoRenewName: computed('config.bmsIsAutoRenew', function() {
+    const bmsIsAutoRenew = get(this, 'config.bmsIsAutoRenew')
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/component.js]: you can destructure here

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs [hunk 8]
```diff
@@ -0,0 +1,667 @@
...
+        </label>
+        <div>{{config.zone}}</div>
+      </div>
+      <div class="col span-6 mb-0">
+        <label class="acc-label">{{t 'clusterNew.huaweicce.projectId.label'}}</label>
+        <div>{{config.projectId}}</div>
+      </div>
+    </div>
+    {{#unless authConfigred}}
+    <div class="row">
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs]: indentation

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs [hunk 9]
```diff
@@ -0,0 +1,667 @@
...
+      <div class="col span-6 mb-0">
+        <label class="acc-label">
+          {{t 'clusterNew.huaweicce.accessKey.label'}}
+        </label>
+        <div>{{config.accessKey}}</div>
+      </div>
+      <div class="col span-6 mb-0">
+        <label class="acc-label">{{t 'clusterNew.huaweicce.secretKey.label'}}</label>
+        <div>**************</div>
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs]: Use the same ui that exists for amazon eks password fields.

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs [hunk 10]
```diff
@@ -0,0 +1,667 @@
...
+            content=containerNetworkCidrContent
+            value=config.containerNetworkCidr
+            allowCustom=true
+          }}
+        {{/if}}
+      </div>
+      <div class="col span-6 mb-0">
+        <label class="acc-label">{{t 'clusterNew.huaweicce.vpcId.label'}}{{field-required}}</label>
+        {{#if editing}}
+          <div class="">
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs]: remove the empty class

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs [hunk 11]
```diff
@@ -0,0 +1,667 @@
...
+        {{/if}}
+      </div>
+    </div>
+    <div class="row">
+      <div class="col span-6 mb-0">
+        <label class="acc-label">
+          {{t 'clusterNew.huaweicce.subnetId.label'}}{{field-required}}
+        </label>
+        {{#if editing}}
+          <div class="">
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs]: remove the empty class

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs [hunk 12]
```diff
@@ -0,0 +1,667 @@
...
+            <div style="margin-top: 10px;">{{t 'clusterNew.huaweicce.subnetId.none'}}</div>
+          {{/if}}
+        {{/if}}
+      </div>
+      {{#if (eq config.clusterType 'BareMetal')}}
+        <div class="col span-6 mb-0">
+          <label class="acc-label">{{t 'clusterNew.huaweicce.highwaySubnet.label'}}</label>
+          {{#if (gt networkContent.length 0)}}
+            {{#if editing}}
+              <div class="">
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs]: remove the empty class

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs [hunk 13]
```diff
@@ -0,0 +1,667 @@
...
+              </div>
+            {{else}}
+              {{new-select
+                classNames="form-control"
+                content=networkContent
+                value=config.highwaySubnet
+              }}
+            {{/if}}
+          {{else}}
+            <div style="margin-top: 10px;">{{t 'clusterNew.huaweicce.highwaySubnet.none'}}</div>
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs]:
```suggestion
            <div class="mt-10">{{t 'clusterNew.huaweicce.highwaySubnet.none'}}</div>
```

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs [hunk 14]
```diff
@@ -0,0 +1,667 @@
...
+          {{/if}}
+        </div>
+      {{/if}}
+    </div>
+    <div class="row">
+      <div class="col span-6 mb-0">
+        <label class="acc-label">{{t 'clusterNew.huaweicce.externalServerEnabled.label'}}</label>
+        {{#if editing}}
+          {{#if config.externalServerEnabled}}
+          <div>{{t 'clusterNew.huaweicce.externalServerEnabled.enabled'}}</div>
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs]: indentation

File: lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs [hunk 15]
```diff
@@ -0,0 +1,667 @@
...
+          {{/if}}
+        </div>
+      {{/if}}
+    </div>
+    <div class="row">
+      <div class="col span-12 mb-0">
+        <label class="acc-label">{{t 'clusterNew.huaweicce.clusterLabels.label'}}</label>
+        {{#if editing}}
+          {{#if (gt config.labels.length 0)}}
+            {{huawei-user-labels readOnly=true
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-huaweicce/template.hbs]: readonly should go to the next line

File: lib/shared/addon/components/cluster-driver/driver-tencentccs/component.js [hunk 16]
```diff
@@ -0,0 +1,530 @@
...
+    label: 'Ubuntu Server 16.04.1 LTS 64bit',
+    value: 'ubuntu16.04.1 LTSx86_64'
+  },
+  {
+    label: 'CentOS 7.2 64bit',
+    value: 'centos7.2x86_64'
+  }
+]
+
+const REGIONS = [{
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-tencentccs/component.js]: formatting

File: lib/shared/addon/components/cluster-driver/driver-tencentccs/component.js [hunk 17]
```diff
@@ -0,0 +1,530 @@
...
+
+    loadNodeConfig(cb) {
+      setProperties(this, { 'errors': null });
+
+      const errors = get(this, 'errors') || [];
+      const intl = get(this, 'intl');
+
+      const clusterCidr = get(this, 'config.clusterCidr');
+      const vpcId = get(this, 'config.vpcId');
+      const goodsNum = get(this, 'config.goodsNum');
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-tencentccs/component.js]: you could destructure here

File: lib/shared/addon/components/cluster-driver/driver-tencentccs/component.js [hunk 18]
```diff
@@ -0,0 +1,530 @@
...
+    },
+
+    loadInstanceConfig(cb) {
+      setProperties(this, { 'errors': null });
+
+      const errors = get(this, 'errors') || [];
+      const intl = get(this, 'intl');
+
+      const zoneId = get(this, 'config.zoneId');
+      const subnetId = get(this, 'config.subnetId');
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-tencentccs/component.js]: you could destructure here

File: lib/shared/addon/components/cluster-driver/driver-tencentccs/component.js [hunk 19]
```diff
@@ -0,0 +1,530 @@
...
+    },
+
+    save(cb) {
+      setProperties(this, { 'errors': null });
+
+      const errors = get(this, 'errors') || [];
+      const intl = get(this, 'intl');
+
+      const sgId = get(this, 'config.sgId');
+      const keyId = get(this, 'config.keyId');
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-tencentccs/component.js]: you could destructure here

File: lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+          }}
+        {{else}}
+          <div>{{config.region}}</div>
+        {{/if}}
+      </div>
+
+      <div class="col span-4">
+        <label class="acc-label">{{t 'clusterNew.tencentccs.secretId.label'}}{{field-required}}</label>
+        {{#if (eq step 1)}}
+          {{input type="text" name="username" classNames="form-control" placeholder=(t 'clusterNew.tencentccs.secretId.placeholder') value=config.secretId}}
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs]: break props on to new lines if over 2

File: lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+        {{else}}
+          <div>{{config.secretId}}</div>
+        {{/if}}
+      </div>
+
+      <div class="col span-4">
+        <label class="acc-label">{{t 'clusterNew.tencentccs.secretKey.label'}}{{field-required}}</label>
+        {{#if (eq step 1)}}
+          {{input type="password" name="password" classNames="form-control" placeholder=(t 'clusterNew.tencentccs.secretKey.placeholder') value=config.secretKey}}
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs]: break props on to new lines if over 2

File: lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs [hunk 22]
```diff
@@ -0,0 +1,293 @@
...
+              value=config.clusterVersion
+            }}
+          {{else}}
+            <div>{{config.clusterVersion}}</div>
+          {{/if}}
+        </div>
+        <div class="col span-6">
+          <label class="acc-label">{{t 'clusterNew.tencentccs.goodsNum.label'}}{{field-required}}</label>
+          {{#if (eq step 2)}}
+            {{input-integer min=1 max=100 value=config.goodsNum classNames="form-control" placeholder=(t 'clusterNew.tencentccs.goodsNum.placeholder')}}
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs]: break props on to new lines if over 2

File: lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs [hunk 23]
```diff
@@ -0,0 +1,293 @@
...
+              value=config.vpcId
+            }}
+          {{else}}
+            <div>{{config.vpcId}}</div>
+          {{/if}}
+        </div>
+        <div class="col span-6">
+          <label class="acc-label">{{t 'clusterNew.tencentccs.cidr.label'}}{{field-required}}</label>
+          {{#if (eq step 2)}}
+            {{input type="text" value=config.clusterCidr placeholder=(t 'clusterNew.tencentccs.cidr.placeholder')}}
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs]: break props on to new lines if over 2

File: lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs [hunk 24]
```diff
@@ -0,0 +1,293 @@
...
+            class="form-control"
+            content=rootDiskChoices
+            localizedLabel=true
+            value=config.rootType
+          }}
+        </div>
+        <div class="col span-6">
+          <label class="acc-label">{{t 'clusterNew.tencentccs.rootSize.label'}}</label>
+          <div class="input-group">
+            {{input-integer min=20 max=50 value=config.rootSize classNames="form-control" placeholder=(t 'clusterNew.tencentccs.rootSize.placeholder')}}
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs]: break props on to new lines if over 2

File: lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs [hunk 25]
```diff
@@ -0,0 +1,293 @@
...
+            class="form-control"
+            content=storageDiskChoices
+            localizedLabel=true
+            value=config.storageType
+          }}
+        </div>
+        <div class="col span-6">
+          <label class="acc-label">{{t 'clusterNew.tencentccs.storageSize.label'}}</label>
+          <div class="input-group">
+            {{input-integer min=0 value=config.storageSize classNames="form-control" placeholder=(t 'clusterNew.tencentccs.storageSize.placeholder')}}
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs]: break props on to new lines if over 2

File: lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs [hunk 26]
```diff
@@ -0,0 +1,293 @@
...
+            class="form-control"
+            content=bandWidthChoices
+            localizedLabel=true
+            value=config.bandwidthType
+          }}
+        </div>
+        <div class="col span-6">
+          <label class="acc-label">{{t 'clusterNew.tencentccs.bandwidth.label'}}</label>
+          <div class="input-group">
+            {{input-integer min=0 value=config.bandwidth classNames="form-control" placeholder=(t 'clusterNew.tencentccs.bandwidth.placeholder')}}
```
Comment (westlywright) [inline lib/shared/addon/components/cluster-driver/driver-tencentccs/template.hbs]: break props on to new lines if over 2

File: lib/shared/addon/components/huawei-user-labels/component.js [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+    let count = this.get('userLabelArray').filterBy('key').get('length') || 0;
+
+    if ( count ) {
+      k = STATUS.COUNTCONFIGURED;
+    }
+
+    this.set('statusClass', classForStatus(k));
+
+    return this.get('intl').t(`${ STATUS_INTL_KEY }.${ k }`, { count });
+  }.property('userLabelArray.@each.key'),
```
Comment (westlywright) [inline lib/shared/addon/components/huawei-user-labels/component.js]: use new computed property syntax

File: lib/shared/addon/components/huawei-user-labels/component.js [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+  updateLabels(labels) {
+    this.sendAction('setLabels', labels);
+    this.validate();
+  },
+
+  validate() {
+    let errors = [];
+
+    //    (this.get('labelArray')||[]).forEach((obj) => {
```
Comment (westlywright) [inline lib/shared/addon/components/huawei-user-labels/component.js]: this looks unfinished.

File: lib/shared/addon/components/huawei-user-labels/template.hbs [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+      <tr class="hidden-xs hidden-sm">
+        <th>{{t 'formUserLabels.key.label'}}{{field-required}}</th>
+        <th width="30">&nbsp;</th>
+        <th>{{t 'formUserLabels.value.label'}}</th>
+        <th width="40">&nbsp;</th>
+      </tr>
+      {{#each userLabelArray as |label|}}
+        <tr>
+          <td data-title="key">
+            {{input-paste pasted="pastedLabels" class="form-control input-sm key" type="text" value=label.key placeholder="formUserLabels.key.placeholder" disabled=readOnly}}
```
Comment (westlywright) [inline lib/shared/addon/components/huawei-user-labels/template.hbs]: break props on to new lines if over 2

File: lib/shared/addon/components/huawei-user-labels/template.hbs [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+          </td>
+
+          <td class="text-center">
+            <p class="input-sm">{{t 'formUserLabels.separator'}}</p>
+          </td>
+
+          <td data-title="label">
+            {{input class="form-control input-sm" type="text" value=label.value placeholder=(t 'formUserLabels.value.placeholder') disabled=readOnly}}
```
Comment (westlywright) [inline lib/shared/addon/components/huawei-user-labels/template.hbs]: break props on to new lines if over 2

**Comment by loganhz** [comment]
@westlywright Updated
Thank you so much for the review. I have fixed the formatting issue in my IDE for `hbs` files

**Pull request merged**